king_phisher/client/windows/campaign_import.py

```diff
@@ -0,0 +1,439 @@
+				continue
+			keys.append(elements.tag)
+			values.append(elements.text)
+
+		self.rpc('db/table/set', 'campaigns', int(self.campaign_info.find('id').text), tuple(keys), tuple(values))
+		nodes_completed += float(len(values) + 1)
+		percentage_completed = nodes_completed / node_count
+		GLib.idle_add(self.import_progress.set_fraction, percentage_completed)
+		if self.thread_import_campaign.stopped():
+			return
+
+		for tables in ('landing_pages', 'messages', 'visits', 'credentials', 'deaddrop_deployments', 'deaddrop_connections'):
+			inserted_ids = []
+			if self.thread_import_campaign.stopped():
+				return
+			GLib.idle_add(self._update_text_view, 'Serializing table {} data for import'.format(tables))
+			keys, rows = self._get_keys_values(self.campaign_info.find(tables))
+			GLib.idle_add(self._update_text_view, 'Working on table {} adding {} rows'.format(tables, len(rows)))
+			if self.thread_import_campaign.stopped():
+				return
+
+			# make table rows easy to manage for updating new ids returned
+			table_rows = []
+			for row in rows:
+				row = dict(zip(keys, row))
+				table_rows.append(row)
+
+			while len(rows) > 0:
+				if self.thread_import_campaign.stopped():
+					return
+				try:
+					inserted_ids = inserted_ids + self.rpc('/db/table/insert/multi', tables, keys, rows[:batch_size], deconflict_ids=True)
+				except advancedhttpserver.RPCError:
+					response = gui_utilities.glib_idle_add_wait(self.failed_import_action)
+					if response:
+						self.remove_import_campaign()
+						GLib.idle_add(self.button_import_campaign.set_sensitive, False)
+						GLib.idle_add(self.select_button.set_sensitive, True)
+						GLib.idle_add(self.spinner.stop)
+						return
+
+				rows = rows[batch_size:]
+				nodes_completed += float(batch_size * len(keys))
+				percentage_completed = nodes_completed / node_count
+				GLib.idle_add(self.import_progress.set_fraction, percentage_completed)
+				if self.thread_import_campaign.stopped():
+					return
+
+			# update id fields to maintain relationships
+			GLib.idle_add(self._update_text_view, 'Updating dependencies for table: {}'.format(tables))
+			for id_ in inserted_ids:
+				if id_ != table_rows[inserted_ids.index(id_)]['id']:
+					self._update_id(
+						self.campaign_info, ['id', '{}_id'.format(tables[:-1])],
+						table_rows[inserted_ids.index(id_)]['id'], id_
+					)
+
+		GLib.idle_add(self.import_progress.set_fraction, 1.0)
+		GLib.idle_add(self.button_import_campaign.set_sensitive, False)
+		self.campaign_info = None
+		GLib.idle_add(self.entry_campaign_name.set_text, '')
+		GLib.idle_add(self.entry_path.set_text, '')
+		GLib.idle_add(self.select_button.set_sensitive, True)
+		GLib.idle_add(self.spinner.stop)
+		done_string = 'Done importing campaign it took {}'.format(datetime.datetime.now() - start_time)
+		GLib.idle_add(self._update_text_view, done_string)
```

Comment:
This paradigm of using `GLib.idle_add` to call `_update_text_view` seems pretty tedious. If there are no instances where you don't call `_update_text_view` from `GLib.idle_add` it would be cleaner to just put the call into the function. If there are, perhaps adding a kwarg of `idle_add` to the function and having it default to what ever is used most would be beneficial.